src/replay/trackers.zig

```diff
@@ -17,27 +17,90 @@ const SlotState = sig.core.SlotState;
 /// kitchen-sink style approach of storing everything under the sun.
 ///
 /// [BankForks](https://github.com/anza-xyz/agave/blob/161fc1965bdb4190aa2d7e36c7c745b4661b10ed/runtime/src/bank_forks.rs#L75)
+///
+/// This struct is *not* thread safe, and the lifetimes of the returned pointers
+/// will end as soon as the items are removed.
 pub const SlotTracker = struct {
-    slots: std.AutoArrayHashMapUnmanaged(Slot, Element) = .{},
+    slots: std.AutoArrayHashMapUnmanaged(Slot, *Element),
+    root: std.atomic.Value(Slot),
 
     const Element = struct {
         constants: SlotConstants,
-        state: SlotState, // TODO properly handle mutations and lifetime
+        state: SlotState,
+    };
+
+    const Reference = struct {
+        constants: *const SlotConstants,
+        state: *SlotState,
     };
 
+    pub fn init(root_slot: Slot) SlotTracker {
+        return .{
+            .slots = .{},
+            .root = .init(root_slot),
+        };
+    }
+
+    pub fn deinit(self: SlotTracker, allocator: Allocator) void {
+        var slots = self.slots;
+        for (slots.values()) |v| allocator.destroy(v);
+        slots.deinit(allocator);
+    }
+
+    pub fn put(
+        self: *SlotTracker,
+        allocator: Allocator,
+        slot: Slot,
+        constants: SlotConstants,
```

Comment:
`SlotConstants` also has a `slot` field. Does that duplicate the `slot` parameter passed in the function? if so, perhaps the slot can be retrieved from `SlotConstants.slot`